prow/cmd/pipeline/controller.go

```diff
@@ -0,0 +1,790 @@
+	logrus.Debugf("createPipelineResource(%s,%s,%s)", context, namespace, pr.Name)
+	pc, err := c.getPipelineConfig(context)
+	if err != nil {
+		return nil, err
+	}
+	return pc.client.TektonV1alpha1().PipelineResources(namespace).Create(pr)
+}
+
+func (c *controller) now() metav1.Time {
+	return metav1.Now()
+}
+
+func (c *controller) pipelineID(pj prowjobv1.ProwJob) (string, error) {
+	if pj.Spec.Refs == nil {
+		return "", fmt.Errorf("no spec refs")
+	}
+	if pj.Spec.Refs.Org == "" {
+		return "", fmt.Errorf("spec refs org is empty")
+	}
+	if pj.Spec.Refs.Repo == "" {
+		return "", fmt.Errorf("spec refs repo is empty")
+	}
+	spec := downwardapi.NewJobSpec(pj.Spec, "", pj.Name)
+	branch := downwardapi.GetBranch(&spec)
+	jobName := fmt.Sprintf("%s/%s/%s", pj.Spec.Refs.Org, pj.Spec.Refs.Repo, branch)
+	return pjutil.GetBuildID(jobName, c.totURL)
+}
+
+var (
+	groupVersionKind = schema.GroupVersionKind{
+		Group:   prowjobv1.SchemeGroupVersion.Group,
+		Version: prowjobv1.SchemeGroupVersion.Version,
+		Kind:    "ProwJob",
+	}
+)
+
+// reconcile ensures a tekton prowjob has a corresponding pipeline, updating the prowjob's status as the pipeline progresses.
+func reconcile(c reconciler, key string) error {
+	logrus.Debugf("reconcile: %s\n", key)
+
+	ctx, namespace, name, kind, err := fromKey(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return nil
+	}
+
+	var wantPipelineRun bool
+	var havePipelineRun bool
+	var reported bool
+	var pj *prowjobv1.ProwJob
+	var p *pipelinev1alpha1.PipelineRun
+	var pr *pipelinev1alpha1.PipelineResource
+
+	switch kind {
+	case prowJob:
+		pj, err = c.getProwJob(name)
+		switch {
+		case apierrors.IsNotFound(err):
+			// Do not want pipeline
+		case err != nil:
+			return fmt.Errorf("get prowjob: %v", err)
+		case pj.Spec.Agent != prowjobv1.TektonAgent:
+			// Do not want a pipeline for this job
+		case pj.Spec.Cluster != ctx:
+			// Build is in wrong cluster, we do not want this build
+			logrus.Warnf("%s found in context %s not %s", key, ctx, pj.Spec.Cluster)
+		case pj.DeletionTimestamp == nil:
+			wantPipelineRun = true
+		}
+
+		if pj == nil {
+			return fmt.Errorf("no prowjob %q found", name)
+		}
+
+		//Add extra check as there is a delay for pipelinerun objects being created and we can get a watch event here
+		//that updates the prowjob with the reporter status and still not have a pipelinerun which causes a duplicate
+		//pipelinerun being requested
+		reported = isReported(&pj.Status)
+
+		selector := fmt.Sprintf("%s = %s", prowJobName, name)
+		p, err = c.getPipelineRunWithSelector(ctx, namespace, selector)
+		switch {
+		case apierrors.IsNotFound(err):
+			// Do not have a pipeline
+		case err != nil:
+			return fmt.Errorf("get pipelinerun %s: %v", key, err)
+		}
+		if p != nil {
+			havePipelineRun = true
+		}
+	case pipelineRun:
+		p, err = c.getPipelineRun(ctx, namespace, name)
```

Comment:
Ideally this should support creating pipelines in foreign clusters, like plank and the build controllers do

Reply:
I can make this multi-cluster aware.

Reply:
The current implementation supports creating pipelines in different clusters. It keeps a map between the cluster context and the pipelineConfig, in a similar way like the build controller. Every time a  ProwJob/PipelineRun is created, it looks up the cluster config based on the given context.

Reply:
Oh I see! Can you choose a different name for the kubectl context than `ctx`? I was thinking this was the `context.Context` builtin.